### PR TITLE
secrecy: rework the `SecretBox` type

### PR DESCRIPTION
**Pull request by fjarri**
Currently `SecretBox = Secret<Box<S>>` is not particularly convenient to use: `Secret<S>` requires `S: Zeroize`, but if one uses a boxed third-party type, it is impossible to `impl Zeroize` for it. `Zeroize` is implemented for `Box<[T]>`, but not for `Box<T>`. 

This PR redefines `SecretBox` as a struct containing `Box<S>`, and impls the same traits for it that `Secret` had.

Changes:
- Removed everything but `SecretBox`. `Secret` cannot be created safely, or even used safely by mutating the contents within the secret (the secret lingers on the stack). `SecretVec/String/Bytes` may be safe if mutation is prohibited, but that just transfers the burden of creating them safely to the user. I suppose we can add them back with a safety note.
- `DebugSecret` removed, `SecretBox` just impls `Debug` now.
- `ExposeSecretMut` added. Not sure about that one, maybe we don't need it as long as `SecretBox::new_with_mut()` is available.
- `SecretBox` can be created: from an existing `Box`, by mutating the contents (`new_with_mut()`), from infallible constructor (`new_with_ctr`), from fallible constructor (`new_with_fallible_ctr`). The names may be changed. The latter method we need for the `Deserialize` impl. I added a safety note to the constructor method indicating that the first two methods are preferable when possible.

Testing notes:
I spent some time trying to introspect the memory to check that no secret data is left on the stack. I used https://docs.rs/psm and https://docs.rs/read-process-memory to check it. It seems that `Secret::new(X::new())` does leave the contents of `X::new()` on the stack, but `SecretBox::new(|| X::new())` (the best I could come up with without resorting to `unsafe`) doesn't. It may be affected by the nature of `X::new()` and Rust version/optimization flags though, but I couldn't come up with a better test. 

It may even be made into an automatic test, if we pick some "safe" memory window to look into without risking a segfault.

### CODE REVIEW

**Comment by fjarri** [comment]
@tarcieri : any comments on this?

File: secrecy/src/boxed.rs
Comment (tony-iqlusion) [inline secrecy/src/boxed.rs:31]: My first thought looking at this is that there has to be another way and that this approach is suboptimal.

I am unfortunately quite aware of how tricky it is to ensure values are constructed on the heap, but this approach seems like it would make ensured inlining more difficult and force the value to be stack allocated, *then* moved to the heap, whereas with proper inlining I believe it's possible to avoid that initial stack allocation.

It's been awhile since I've checked that in godbolt though, and I think work on placement-by-return has largely stalled (and the `box` keyword removed).

Comment (tony-iqlusion) [reply secrecy/src/boxed.rs:31]: A simpler alternative is to accept `Box<T>` as an argument.

I think a complex constructor like this needs to be informed by and justified by the generated assembly. If it leaves any copies on the stack it defeats the point (and potentially the inliner).

Comment (fjarri) [reply secrecy/src/boxed.rs:31]: > I am unfortunately quite aware of how tricky it is to ensure values are constructed on the heap, but this approach seems like it would make ensured inlining more difficult and force the value to be stack allocated, then moved to the heap, whereas with proper inlining I believe it's possible to avoid that initial stack allocation.

I don't think it is supported in Rust currently. It's called "placement new", and currently only exists [as a proposal](https://github.com/PoignardAzur/placement-by-return/blob/placement-by-return/text/0000-placement-by-return.md). At the moment we have to accept the fact the we have to construct the value on stack, and all we can do is at least contain it within a small closure and zeroize it. 

> A simpler alternative is to accept Box<T> as an argument.

That is certainly a good idea as an additional constructor, but it's not always possible to already have a `Box`.

Comment (tony-iqlusion) [reply secrecy/src/boxed.rs:31]: > I don't think it is supported in Rust currently. It's called "placement new", and currently only exists [as a proposal](https://github.com/PoignardAzur/placement-by-return/blob/placement-by-return/text/0000-placement-by-return.md).

You just linked to the placement-by-return proposal I was referring to in my comments.

But without placement-by-return, you're going to make a copy when `ctr` returns a value, which defeats the whole point.

The only way to eliminate the copy is by eliminating the intermediate stack frame, i.e. inlining, which it seems like this approach will always defeat.

Comment (tony-iqlusion) [reply secrecy/src/boxed.rs:31]: In other words, this constructor needlessly forces a stack allocation of a value, *then* copies it to the heap.

I think we should avoid making copies as much as possible.

Comment (tony-iqlusion) [reply secrecy/src/boxed.rs:31]: > Not according to my tests.

What tests did you do?

**Comment by tony-iqlusion** [comment]
The rest of this PR is great and the direction I wanted to go with `secrecy`.

If you can provide a simpler constructor, e.g. one which accepts `Box` as an argument, we can get it merged and then perhaps circle back on a more complex constructor.

**Comment by fjarri** [comment]
Any comments on the rest of the listed things in the starting post? Or should they be left for other PRs?

**Comment by tony-iqlusion** [comment]
- If you just accept a `Box<T>` in the constructor you don't need a `Clone` bound
- The point of `DebugSecret` was to prevent leakage of secret values. I'd personally prefer to retain it. Without it, it becomes very easy to log secrets. If the type doesn't impl `DebugSecret`, they lose the `Debug` impl. If they really want to log the secret, they can access the secret value.
- The reason for an inherent accessor method was to make secret accesses easy to search for. `AsRef`/`AsMut` defeat that
- I'm not sure there's enough functionality to warrant decomposing into submodules

File: secrecy/src/boxed.rs
Comment (fjarri) [reply secrecy/src/boxed.rs:31]: > In other words, this constructor needlessly forces a stack allocation of a value, then copies it to the heap.

Well, either you have a `Box::new(T::default())` and then mutate it, or you construct it on the stack. The former is not always possible or convenient, and the latter seems to be safe with the constructor I provided.

> What tests did you do?

Creating `Secret` values in a function and inspecting the stack with `psm` and `read-process-memory` I linked in the description. I'm thinking on how to make it into an autotest, since it's a little flaky.

Comment (tony-iqlusion) [reply secrecy/src/boxed.rs:31]: I think your proposed constructor would need a lot of testing on different platforms to determine it accomplishes the desired goals, yes. Can we hold off on it for now?

Comment (tarcieri) [reply secrecy/src/boxed.rs:31]: The `Clone` bound is another argument: since the only constructor requires a `Clone` bound, it means all your secret types *MUST* impl `Clone`, which may be undesirable.

**Comment by fjarri** [comment]
>  If the type doesn't impl DebugSecret, they lose the Debug impl.

Yes, but what if I want a `Debug` impl (which masks the secret value) for a `SecretBox<T>` where `T` is a third-party type? I can't implement `Debug` on `SecretBox`, and I can't implement `DebugSecret` on `T`.  It seems to me that leakage can just as easily be prevented by simply providing `impl Debug for SecretBox` using `any`, which is what I did.

> The reason for an inherent method as an accessor method was to make secret accesses easy to search for. AsRef/AsMut defeat that, and combined with functions that bound on AsRef/AsMut and call the relevant trait methods internally would make secret accesses very difficult to locate.

I was thinking that it would simplify using secret values in generic context, but I see your point.

> I'm not sure there's enough functionality to warrant decomposing into submodules

If we remove everything but `SecretBox` then certainly. But currently there are submodules.

File: secrecy/src/boxed.rs
Comment (fjarri) [reply secrecy/src/boxed.rs:31]: We can if we must, but I'll probably still have to use it in my own code since I cannot create a random `crypto_bigint::Uint` or `k256::Scalar` directly in a `Box`.

**Comment by tarcieri** [comment]
Okay, your `Debug` impl is fine. `type_name` wasn't available at the time `secrecy` was written.

> If we remove everything but SecretBox then certainly. But currently there are submodules.

I think we should get rid of all of the types besides `SecretBox`.

`SecretString` can be `pub type SecretString = SecretBox<str>`.

`SecretVec` can be replaced with `SecretBox<[T]>` (which could potentially have a `SecretSlice` alias if helpful).

**Comment by tony-iqlusion** [comment]
Okay, your `Debug` impl is fine. `type_name` wasn't available at the time `secrecy` was written.

> If we remove everything but SecretBox then certainly. But currently there are submodules.

I think we should get rid of all of the types besides `SecretBox`.

`SecretString` can be `pub type SecretString = SecretBox<str>` (or perhaps `SecretStr`).

`SecretVec` can be replaced with `SecretBox<[T]>` (which could potentially have a `SecretSlice` alias if helpful).

**Comment by fjarri** [comment]
Also, 

> If you just accept a Box<T> in the constructor you don't need a Clone bound

That's not a problem, we can have a separate `impl` block bound on `Clone` for that specific constructor, without requiring it for the rest of the methods.

**Comment by tony-iqlusion** [comment]
@fjarri but it's currently the *only* constructor, leading to a de facto `Clone` bound for the whole type.

If you had `pub fn new(secret: Box<T>)` then that's no longer an issue.

You can then add a second constructor that does your `FnOnce`-based allocator with a mandatory `Clone` bound.

**Comment by fjarri** [comment]
Oh yes, I have no problems with having a constructor that takes `Box` (or even a `From` impl?). I actually wanted to add it myself, but I must have forgotten to do it.

**Comment by tony-iqlusion** [comment]
A `From` impl is fine as long as orphan rules allow it to be sufficiently generic

**Comment by tony-iqlusion** [comment]
> either you have a Box::new(T::default()) and then mutate it

This potentially seems like a better pattern worth supporting, i.e. a constructor which handles the allocation using `Default`, then passes `&mut T` to a callback function for initialization.

**Comment by fjarri** [comment]
Removed all the stuff, updated the PR description accordingly.

File: secrecy/src/lib.rs
Comment (tony-iqlusion) [inline secrecy/src/lib.rs:119]: I think this is a bit more idiomatic:

```suggestion
    pub fn try_new_with_ctr<E>(ctr: impl FnOnce() -> Result<S, E>) -> Result<Self, E> {
```

File: secrecy/src/lib.rs
Comment (tony-iqlusion) [inline secrecy/src/lib.rs:141]:
```suggestion
        write!(f, "SecretBox<{}>([REDACTED])", any::type_name::<S>())
```

Comment (fjarri) [reply secrecy/src/lib.rs]: Fixed.

File: secrecy/src/lib.rs
Comment (fjarri) [reply secrecy/src/lib.rs]: Fixed.

**Comment by tyranron** [comment]
Any new on this?

**Comment by tony-iqlusion** [comment]
I'd be happy to merge it if it weren't still marked a draft

**Review by tony-iqlusion — APPROVED** [review]

**Pull request merged**
